File: 1140-stone-game-ii/1140-stone-game-ii.java
class Solution {
    public int n;
    public int[][][] dp;
    public int[] sum;
    
    public int getSum(int start, int end) {        
        int totalValue = sum[n - 1];
        int startValue = start - 1 < 0 ? 0 : sum[start - 1];
        int endValue = sum[Math.min(n - 1, end)];    
        int result = (sum[n - 1] - startValue) - (sum[n - 1] - endValue);        
        return result;
    }
    
    public int dfs(int alice, int x, int index) {
        if (n <= index) {
            return 0;
        }
        
        if (dp[alice][x][index] != 0) {
            return getSum(index, n - 1) - dp[alice][x][index];
        }
            
        for (int i = 1; i <= x * 2; i++) {        
            int v = getSum(index, index + i - 1);            
            v += dfs((alice + 1) % 2, Math.max(i, x), index + i);
            dp[alice][x][index] = Math.max(dp[alice][x][index], v);
        }        
        
        return getSum(index, n - 1) - dp[alice][x][index];
    }
    
    public int stoneGameII(int[] piles) {
        n = piles.length;
        dp = new int[2][n * 2][n];
        sum = new int[n];
        sum[0] = piles[0];
        
        for (int i = 1; i < piles.length; i++) {        
            sum[i] = piles[i] + sum[i - 1];
        }
        
        dfs(0, 1, 0);
        return dp[0][1][0];
    }
}

File: 1140-stone-game-ii/1140-stone-game-ii.cpp
class Solution {
public:
    typedef vector<int>vi;
    typedef vector<vi>vii;
    typedef vector<vii>viii;
    viii dp;
    vi p;
    int n;
    vi sum;
    int dfs(int turn, int idx,int m){
        if(idx==n){
            return 0;
        }
        if(dp[turn][idx][m]!=1e9){
            return dp[turn][idx][m];
        }
        for(int i=1;i<=2*m;i++){
            if(n<idx+i){
                break;
            }
            int s=sum[idx+i-1]-((idx==0)?0:sum[idx-1]);
            if(turn==0){
                if(dp[turn][idx][m]==1e9){
                    dp[turn][idx][m]=dfs(1,idx+i,max(i,m))+s;
                }else{
                    dp[turn][idx][m]=max(dp[turn][idx][m],dfs(1,idx+i,max(i,m))+s);
                }
            }else{
                dp[turn][idx][m]=min(dp[turn][idx][m],dfs(0,idx+i,max(i,m)));
            }
        }
        return dp[turn][idx][m];
    }
    int stoneGameII(vector<int>& piles) {
        p=piles;
        n=piles.size();
        dp=viii(2,vii(n+1,vi(n+1,1e9)));
        sum=vi(n,p[0]);
        for(int i=1;i<n;i++){
            sum[i]=sum[i-1]+p[i];
        }
        int result= dfs(0,0,1);
        return result;
    }
};